src/Subscriber/OrderStateChangeSubscriber.php

```diff
@@ -0,0 +1,146 @@
+<?php
+
+namespace Wexo\AltaPay\Subscriber;
+
+use Exception;
+use Psr\Log\LoggerInterface;
+use SimpleXMLElement;
+use Shopware\Core\Checkout\Order\OrderEntity;
+use Shopware\Core\Checkout\Order\Event\OrderStateMachineStateChangeEvent;
+use Shopware\Core\Checkout\Order\Aggregate\OrderTransaction\OrderTransactionStateHandler;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Wexo\AltaPay\Service\PaymentService;
+
+class OrderStateChangeSubscriber implements EventSubscriberInterface
+{
+
+
+    public function __construct(
+        protected readonly PaymentService               $paymentService,
+        protected readonly OrderTransactionStateHandler $orderTransactionStateHandler,
+        protected readonly LoggerInterface              $logger,
+    )
+    {
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            'state_enter.order.state.completed' => 'onOrderStateComplete',
+            'state_enter.order_delivery.state.returned' => 'onOrderDeliveryStateReturned',
+        ];
+    }
+
+    public function onOrderStateComplete(OrderStateMachineStateChangeEvent $event): void
+    {
+        try {
+            $context = $event->getContext();
+            $order = $event->getOrder();
+
+            if (!$order || empty($order->getCustomFields()[PaymentService::ALTAPAY_TRANSACTION_ID_CUSTOM_FIELD] ?? null)) {
+                return;
+            }
+
+            $altaPayTransaction = $this->getAltaPayTransaction($order);
+
+            if ($altaPayTransaction && (float)$altaPayTransaction->ReservedAmount > 0.0) {
+
+                $orderTotal = $order->getAmountTotal();
+                $capturedAmount = (float)$altaPayTransaction->CapturedAmount;
+
+                // Calculate the remaining amount that can be captured
+                $remainingAmount = $orderTotal - $capturedAmount;
+
+                // Ensure the remaining amount is not negative and less than or equal to the reserved amount
+                if ($remainingAmount > 0.0 && $remainingAmount <= (float)$altaPayTransaction->ReservedAmount) {
+
+                    $response = $this->paymentService->captureReservation($order, $order->getSalesChannelId(), $remainingAmount);
+                    $responseAsXml = new SimpleXMLElement($response->getBody()->getContents());
+
+                    if ((string)$responseAsXml->Body?->Result === "Success") {
+                        $this->orderTransactionStateHandler->paid(
+                            $order->getTransactions()->first()->getId(),
+                            $context
+                        );
+                    }
```

Comment:
What if it is not successful? Shouldn't we log the error?

Reply:
@emicha , I’ve added the log.